ocp_utilities/data_collector.py

```diff
@@ -0,0 +1,173 @@
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+def collect_vmi_data(vmi, directory=None):
+    write_to_file(
+        file_name=f"vmi-{vmi.name}.yaml",
+        content=vmi.instance.to_str(),
+        base_directory=directory,
+    )
+
+    virt_launcher_pod = vmi.virt_launcher_pod
+    write_to_file(
+        file_name=f"{virt_launcher_pod.name}.log",
+        content=virt_launcher_pod.log(container="compute"),
+        base_directory=directory,
+    )
+    write_to_file(
+        file_name=f"{virt_launcher_pod.name}.yaml",
+        content=virt_launcher_pod.instance.to_str(),
+        base_directory=directory,
+    )
+
+
+def collect_data_volume_data(dyn_client, directory=None):
+    cdi_pod_prefixes = ("importer", "cdi-upload")
+    for pod in Pod.get(dyn_client=dyn_client):
+        pod_name = pod.name
+        if pod_name.startswith(cdi_pod_prefixes) or pod_name.endswith("source-pod"):
+            write_to_file(
+                file_name=f"{pod_name}.log", content=pod.log(), base_directory=directory
+            )
+            write_to_file(
+                file_name=f"{pod_name}.yaml",
+                content=pod.instance.to_str(),
+                base_directory=directory,
+            )
+
+
+def collect_data(directory, resource_object, collect_pod_logs):
+    LOGGER.info(f"Collecting instance data for {resource_object.name}")
+
+    if resource_object.kind == Pod.kind:
+        collect_pods_data(
+            pods_list=[resource_object],
+            collect_pod_logs=collect_pod_logs,
+            base_directory=directory,
+        )
+        return
+
+    if resource_object.kind == ProjectRequest.kind:
+        resource_object = Namespace(name=resource_object.name)
+
+    write_to_file(
+        file_name=f"{resource_object.name}.yaml",
+        content=resource_object.instance.to_str(),
+        base_directory=directory,
+    )
+    if resource_object.kind == VirtualMachine.kind and resource_object.vmi:
+        collect_vmi_data(vmi=resource_object.vmi, directory=directory)
+
+    if resource_object.kind == DataVolume.kind:
+        collect_data_volume_data(dyn_client=resource_object.client, directory=directory)
+
+
+def prepare_test_data_dir(item, prefix, logs_path=None):
+    test_cls_name = item.cls.__name__ if item.cls else ""
+    test_dir_log = os.path.join(
+        logs_path or py_config["data_collector"]["data_collector_base_directory"],
+        item.fspath.dirname.split("/tests/")[-1],
+        item.fspath.basename.partition(".py")[0],
+        test_cls_name,
+        item.name,
+        prefix,
+    )
+    os.makedirs(test_dir_log, exist_ok=True)
+    return test_dir_log
+
+
+def collect_resources_yaml_instance(resources_to_collect, namespace_name=None):
+    get_kwargs = {"dyn_client": get_client()}
+    for _resources in resources_to_collect:
+        try:
+            if isinstance(_resources, NamespacedResource):
+                get_kwargs["namespace"] = namespace_name
```

Comment:
if `get_kwargs` is initialized outside the for loop, would not this impact data collection if the list contains [namespaced-resource, resource]? It won't be able to find resource via get call.

Reply:
Getting a cluster scoped resource will not fail:
```
In [11]: get_kwargs
Out[11]: 
{'dyn_client': <openshift.dynamic.client.DynamicClient at 0x7f137856fc70>,
 'namespace': 'default'}

In [12]: list(Node.get(**get_kwargs))[0]
Out[12]: <ocp_resources.node.Node at 0x7f1372584130>

In [13]: list(Node.get(**get_kwargs))[0].name
Out[13]: 'ip-10-0-142-209.ec2.internal'
```

Reply:
However, I can move `get_kwargs` inside the loop not to abuse kwargs; wdyt?

Reply:
👍 I like the idea of moving `get_kwargs` inside the loop. Thank you!